### PR TITLE
Refactor tokens to use NSTextAttachment

### LINKED ISSUE #666: Fix tag displaying
Currently tag displaying in the start time entry view is broken if the tags span multiple lines.

screen shot 2017-10-03 at 11 33 55

### LINKED ISSUE #890: Selecting text not working properly
Both the start view text field and the edit view description text field (maybe others as well?) behave strangely with selections. Sometimes they just deselect or don't allow me to select the letters I want, instead selecting whole words or the entire text. It seems very finicky and I'm not sure what the pattern is, but it doesn't seem right.

### PR DESCRIPTION
**Pull request by willsb**
Closes #666 
Closes #890 

🐉 🐉 BEAST SLAYER 🐉 🐉

Do a thorough functional test, pretty please.

### CODE REVIEW

**Comment by arturs105** [comment]
Did a functional review:
- When tapping and holding a tag i can save it as an image nov-23-2017 12-47-41
- Having a very long project name does this (see gif). Though i'm not sure whether we can/should do something about this nov-23-2017 12-50-18
- When there is enough text for scrolling, tapping on the text to move the cursor, srcolls to the bottom. There are also some scroll glitches when typing some stuff in a very long description.

**Comment by pe0ny** [comment]
- What happens when you just tap on it - it *should* open a project/tags view depending on if you tap on a project or tag token?
- Long project name looks nice. I think it makes sense to keep it that way with dots at the end.

File: Toggl.Daneel/Binding/TextViewTextInfoTargetBinding.cs [hunk 1]
```diff
@@ -73,80 +73,37 @@ private void onSelectedTextRangeChanged(NSObservedChange change)
...
-                        .WithTextAndCursor(description, tagStart));
-                return;
-            }
+            onTextFieldInfoChanged(
+                textFieldInfo
+                    .RemoveTag(textFieldInfo.Tags[e.TagIndex])
+                    .WithTextAndCursor(textFieldInfo.Text, e.CursorPosition));
         }
 
+        private bool isSelectingText;
```
Comment (arturs105) [inline Toggl.Daneel/Binding/TextViewTextInfoTargetBinding.cs]: Any reason this isn't defined at the top of the class?

Comment (willsb) [reply Toggl.Daneel/Binding/TextViewTextInfoTargetBinding.cs]: No, this is wrong

**Comment by willsb** [comment]
I think what Arturs meant was that the cursor is on the second line. And yes, there isn't much we can do except truncating the name a little earlier. Right now we truncate at 50 chars. What do you think, @pe0ny ?

**Comment by pe0ny** [comment]
We can keep it at 50 chars. Users tend to use long project/client/tasks/tag names.

File: Toggl.Daneel/Extensions/TokenExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+        private static UIImage getProjectToken(NSAttributedString stringToDraw, UIColor projectColor)
+        {
+            const int circleWidth = circleDiameter + circlePadding;
+            var totalWidth = stringToDraw.Size.Width + circleWidth + interTokenSpacing + (tokenPadding * 2);
+            var size = new CGSize(totalWidth, lineHeight);
+            UIGraphics.BeginImageContextWithOptions(size, false, 0.0f);
+            using (var context = UIGraphics.GetCurrentContext())
+            {
+                var tokenPath = UIBezierPath.FromRoundedRect(new CGRect(
+                    x: interTokenSpacing, y: tokenVerticallOffset,
```
Comment (arturs105) [inline Toggl.Daneel/Extensions/TokenExtensions.cs]: I'd prefer these to be on separate lines.

File: Toggl.Daneel/Extensions/TokenExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+                    width: totalWidth - interTokenSpacing,
+                    height: tokenHeight
+                ), tokenCornerRadius);
+                context.AddPath(tokenPath.CGPath);
+                context.SetFillColor(projectColor.ColorWithAlpha(0.12f).CGColor);
+                context.FillPath();
+
+                var circle = UIBezierPath.FromRoundedRect(new CGRect(
+                    x: circlePadding + interTokenSpacing, y: circleYOffset,
```
Comment (arturs105) [inline Toggl.Daneel/Extensions/TokenExtensions.cs]: Same

File: Toggl.Daneel/Extensions/TokenExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+                var tokenPath = UIBezierPath.FromRoundedRect(new CGRect(
+                    x: interTokenSpacing, y: tokenVerticallOffset,
+                    width: totalWidth - interTokenSpacing,
+                    height: tokenHeight
+                ), tokenCornerRadius);
+                context.AddPath(tokenPath.CGPath);
+                context.SetFillColor(projectColor.ColorWithAlpha(0.12f).CGColor);
+                context.FillPath();
+
+                var circle = UIBezierPath.FromRoundedRect(new CGRect(
```
Comment (arturs105) [inline Toggl.Daneel/Extensions/TokenExtensions.cs]: Is this the small project dot? If so, i think `dot` or `projectDot` would be a better name.

File: Toggl.Daneel/Extensions/TokenExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+        }
+
+        private static UIImage getTagToken(NSAttributedString stringToDraw)
+        {
+            var size = new CGSize(stringToDraw.Size.Width + interTokenSpacing + (tokenPadding * 2), lineHeight);
+            UIGraphics.BeginImageContextWithOptions(size, false, 0.0f);
+            using (var context = UIGraphics.GetCurrentContext())
+            {
+                var tokenPath = UIBezierPath.FromRoundedRect(new CGRect(
+                    x: interTokenSpacing, y: tokenVerticallOffset,
```
Comment (arturs105) [inline Toggl.Daneel/Extensions/TokenExtensions.cs]: You know

File: Toggl.Daneel/TokenTextAttachment.cs [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+using CoreGraphics;
+using UIKit;
+
+namespace Toggl.Daneel
+{
+    public sealed class TokenTextAttachment : NSTextAttachment
+    {
+        public nfloat FontDescender { get; set; }
+
+        public override CGRect GetAttachmentBounds(NSTextContainer textContainer, CGRect proposedLineFragment, CGPoint glyphPosition, nuint characterIndex)
```
Comment (arturs105) [inline Toggl.Daneel/TokenTextAttachment.cs]: This line (and maybe 13. line as well) should be broken into multiple lines.

File: Toggl.Daneel/Extensions/TokenExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+            };
+
+            textVerticalOffset = (lineHeight / 2) - (tokenFont.CapHeight / 2) - 3;
+        }
+
+        public static string GetDescription(this UITextView self) => self.Text.Replace("￼", "");
+
+        public static NSAttributedString GetAttributedText(this TextFieldInfo self)
+        {
+            var result = new NSMutableAttributedString(self.Text);
```
Comment (Odrakir) [inline Toggl.Daneel/Extensions/TokenExtensions.cs]: Not a big deal, but there's an initializer in which you send the text and the attributes, so you don't need to add them afterwards with the Range calculation and all that.

File: Toggl.Daneel/Extensions/TokenExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+
+                stringToDraw.DrawString(new CGPoint(circleWidth + interTokenSpacing + tokenPadding, textVerticalOffset));
+
+                var image = UIGraphics.GetImageFromCurrentImageContext();
+                UIGraphics.EndImageContext();
+                return image;
+            }
+        }
+
+        private static UIImage getTagToken(NSAttributedString stringToDraw)
```
Comment (Odrakir) [inline Toggl.Daneel/Extensions/TokenExtensions.cs]: Couldn't you move this and the previous method to initializeres on TokenTextAttachment, so this class is cleaner?

File: Toggl.Daneel/Extensions/TokenExtensions.cs [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+using Toggl.Multivac.Extensions;
+using Toggl.Foundation.Autocomplete;
+using Toggl.Foundation.MvvmCross.Helper;
+using UIKit;
+
+namespace Toggl.Daneel.Extensions
+{
+    public static class TokenExtensions
+    {
+        private const int lineHeight = 24;
```
Comment (Odrakir) [inline Toggl.Daneel/Extensions/TokenExtensions.cs]: You could probably get lineHeight from the attributed string.

**Comment by arturs105** [comment]
This still has some scrolling issues.

nov-29-2017 11-20-00

**Comment by willsb** [comment]
@arturs105 I think this is best left for another PR, when we refactor it to allow text on both side. Then the user will always be editing the end of the text and this will have appropriate behavior. What do you say?

**Comment by arturs105** [comment]
Yeah, i was kinda expecting that answer.

**Pull request merged**